src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java

```diff
@@ -0,0 +1,136 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis.actions;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.actions.Action;
+import com.google.devtools.build.lib.actions.ActionKeyContext;
+import com.google.devtools.build.lib.actions.Artifact.ArtifactExpander;
+import com.google.devtools.build.lib.actions.CommandLineExpansionException;
+import com.google.devtools.build.lib.actions.CommandLines.CommandLineLimits;
+import com.google.devtools.build.lib.actions.PathMapper;
+import com.google.devtools.build.lib.actions.Spawn;
+import com.google.devtools.build.lib.analysis.config.BuildConfigurationValue;
+import com.google.devtools.build.lib.analysis.config.CoreOptions;
+import com.google.devtools.build.lib.analysis.config.CoreOptions.OutputPathsMode;
+import com.google.devtools.build.lib.util.Fingerprint;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import javax.annotation.Nullable;
+
+/**
+ * Utility methods that are the canonical way for actions to support path mapping (see {@link
+ * PathMapper}).
+ */
+public final class PathMappers {
+  // TODO: Replace with a command-line flag.
+  private static final ImmutableSet<String> SUPPORTED_MNEMONICS = ImmutableSet.of(
+      "AndroidLint",
+      "CompileAndroidResources",
+      "DeJetify",
+      "DejetifySrcs",
+      "Desugar",
+      "DexBuilder",
+      "Javac",
+      "JavacTurbine",
+      "Jetify",
+      "JetifySrcs",
+      "LinkAndroidResources",
+      "MergeAndroidAssets",
+      "MergeManifests",
+      "ParseAndroidResources",
+      "StarlarkAARGenerator",
+      "StarlarkMergeCompiledAndroidResources",
+      "StarlarkRClassGenerator",
+      "Turbine"
+  );
+
+  /**
+   * Actions that support path mapping should call this method from {@link
+   * Action#getKey(ActionKeyContext, ArtifactExpander)}.
+   *
+   * <p>Compared to {@link #create(Action, OutputPathsMode)}, this method does not flatten nested
+   * sets and thus can't result in memory regressions.
+   *
+   * @param mnemonic the mnemonic of the action
+   * @param outputPathsMode the value of {@link CoreOptions#outputPathsMode}
+   * @param fingerprint the fingerprint to add to
+   */
+  public static void addToFingerprint(String mnemonic, OutputPathsMode outputPathsMode,
+      Fingerprint fingerprint)
+      throws CommandLineExpansionException, InterruptedException {
+    // Creating a new PathMapper instance can be expensive, but isn't needed here: Whether and
+    // how path mapping applies to the action only depends on the output paths mode and the action
+    // inputs, which are already part of the action key.
+    OutputPathsMode effectiveOutputPathsMode = getEffectiveOutputPathsMode(outputPathsMode,
+        mnemonic);
+    if (effectiveOutputPathsMode == OutputPathsMode.STRIP) {
+      fingerprint.addString(StrippingPathMapper.GUID);
```

Comment:
Why this GUID specifically? Isn't the PathMappers interface more general than just StrippingPathMapper?